python/handleInvalidState.py

```diff
@@ -92,10 +94,36 @@ def checkGoalValidity(state):
     Returns:
        bool True iff there is no obstacle projected to be on the state's goal waypoint
     '''
-    referenceLatlon = state.globalWaypoint
+    if goalLatlon is None:
+        goalLatlon = state.globalWaypoint
+
+    referenceLatlon = goalLatlon
     obstacles = obs.getObstacles(state, referenceLatlon)
     for obstacle in obstacles:
-        goalXY = utils.latlonToXY(state.globalWaypoint, referenceLatlon)
+        goalXY = utils.latlonToXY(goalLatlon, referenceLatlon)
         if not obstacle.isValid(goalXY):
             return False
     return True
+
+
+def getNewGlobalWaypoint(state):
```

Comment:
The function implementation looks really good to me, but a few comments:
1. Consider changing name to "getValidGlobalWaypoint" or "getClosestValidGlobalWaypoint" or "moveGlobalWaypointUntilValid", to make it clear it is not getting the NEXT waypoint, but getting the closest latlon to the global waypoint that is valid
2. Docstring is good, but more detail would be nice. Keep what you have. But also add something like "moves the global waypoint toward/away from sailbot until valid. Tbh just reading the code quickly, I'm not 100% sure if it moves TOWARD or AWAY from the boat, so that would be really helpful. Otherwise the code is very clean and easy to understand.
3. In this code, MOVE_GOAL_WAYPOINT_STEP_SIZE is very clear. But at the top of file, MOVE_GOAL_WAYPOINT_STEP_SIZE is a bit unclear what unit this is (unitless or km etc.). Consider adding _KM at the end